### PR TITLE
LF-3545, LF-3634 Viewing the task list after another user has created new crop plans causes whitescreen type error crash

### PR DESCRIPTION
**Pull request by kathyavini**
**Description**

This undoes several code changes that were implemented to allow tasks to handle hard-deleted (in the Redux store) management plans, which nonetheless created a chain effect of type error bugs.

As the management plan entities are too intertwined with the task entities (which have a very particular order of loading as well), I moved management plan deletion to to soft delete _on the frontend_, filtered in the main selector (but not the entities selector), and reverted the changes to task.

Jira link: https://lite-farm.atlassian.net/browse/LF-3545

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

I tested this with two users logged onto the same farm, looking at the task list after the other user changed (added/deleted) management plans.

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/webapp/src/containers/Task/index.jsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ import {
   userFarmSelector,
 } from '../userFarmSlice';
 import { resetAndUnLockFormData } from '../hooks/useHookFormPersist/hookFormPersistSlice';
-import { getCropsAndManagementPlans } from '../saga';
+import { getManagementPlansAndTasks } from '../saga';
```
Comment (kathyavini) [inline packages/webapp/src/containers/Task/index.jsx:30]: The changes in this file are directly undoing the changes from this PR:

- https://github.com/LiteFarmOrg/LiteFarm/pull/2829

File: packages/api/src/controllers/taskController.js [hunk 2]
```diff
@@ -679,7 +679,7 @@ const taskController = {
         .whereNotDeleted()
         .withGraphFetched(
           `[locations.[location_defaults], managementPlans, soil_amendment_task, field_work_task.[field_work_task_type], cleaning_task, pest_control_task, 
-            harvest_task.[harvest_use], plant_task, transplant_task, irrigation_task.[irrigation_type]]
+            harvest_task.[harvest_use], plant_task.[planting_management_plan.[bed_method, container_method, broadcast_method, row_method]], transplant_task, irrigation_task.[irrigation_type]]
```
Comment (kathyavini) [inline packages/api/src/controllers/taskController.js:682]: This change doesn't actually impact on this specific bug.

However adding this part of the graph was based on `getPlantingTasksAndPlantingManagementPlansSuccessSaga` being set up to get and then forward the `planting_management_property` (which had previously always just not existed). There may be another task + plans bug that is therefore addressed by this; at the least it shouldn't hurt as it's not difficult to fetch this extra part of the graph.

File: packages/webapp/src/containers/managementPlanSlice.js [hunk 3]
```diff
@@ -61,6 +62,22 @@ const addManyManagementPlan = (state, { payload: managementPlans }) => {
   );
 };
 
+const deleteOneManagementPlan = (state, { payload }) => {
```
Comment (JimmyRowland) [inline packages/webapp/src/containers/managementPlanSlice.js:65]: {payload : management_plan_id}

File: packages/webapp/src/containers/Task/index.jsx [hunk 4]
```diff
@@ -99,13 +99,10 @@ export default function TaskPage({ history }) {
   }, [activeUsers.length, taskTypes.length]);
 
   useEffect(() => {
-    dispatch(getTasks());
     dispatch(getTaskTypes());
```
Comment (JimmyRowland) [inline packages/webapp/src/containers/Task/index.jsx:102]: getManagementPlansAndTasks calls getTaskTypes.
getTaskTypes can be removed

File: packages/webapp/src/containers/managementPlanSlice.js [hunk 5]
```diff
@@ -61,6 +62,22 @@ const addManyManagementPlan = (state, { payload: managementPlans }) => {
...
+  const managementPlan = state.entities[payload];
+  if (managementPlan) {
+    managementPlan.deleted = true;
+  }
+};
+
+const deleteManyManagementPlans = (state, { payload }) => {
+  payload.forEach((id) => {
```
Comment (JimmyRowland) [inline packages/webapp/src/containers/managementPlanSlice.js:73]: {payload : managementPlanIds}

**Review by JimmyRowland — APPROVED** [review]

**Comment by JimmyRowland** [comment]
I think there should be a follow up PR
In the management plan selector, deleted ones can be filtered out.
image

In tasksSelector, complemented/abandoned tasks can be filtered out if all locations and management plans are deleted.

After, a proper fix can be implemented. Whether it's WebSocket or cache invalidation on each fetch

**Comment by kathyavini** [comment]
Thank you for the suggestions @JimmyRowland!! For some reason putting the deleted check in `tasksSelector` didn't work -- it looked like maybe the deleted management plan wasn't in the task object at that point? -- and I gotta admit my comprehension of `taskSlice` is not 100% so I couldn't get to the bottom of that.

That said, filtering on the task cards (in `taskCardsContentSelector`) works and seems to hit all the UI places (including locations) where the inconsistency could be spotted so I'm hoping that is enough! 🤞

**Comment by JimmyRowland** [comment]
does this work? `export const tasksSelector = createSelector(
  [taskEntitiesSelector, loginSelector],
  (taskEntities, { farm_id }) => {
    return Object.values(taskEntities).filter(({ locations, managementPlans, taskType }) => {
      if (
        (locations.length || managementPlans.length) &&
        [locations, managementPlans].reduce(
          (allDeleted, array) => allDeleted && array.every(({ deleted }) => deleted),
          true,
        )
      ) {
        return false;
      }
      for (const location of locations) {
        if (location.farm_id === farm_id) {
          return true;
        }
      }
      for (const managementPlan of managementPlans) {
        if (managementPlan.farm_id === farm_id) {
          return true;
        }
      }
      if (taskType.farm_id === farm_id) {
        return true;
      }
      return false;
    });
  },
);`

**Review by JimmyRowland — DISMISSED** [review]

File: packages/webapp/src/containers/Task/index.jsx [hunk 6]
```diff
@@ -99,13 +98,9 @@ export default function TaskPage({ history }) {
   }, [activeUsers.length, taskTypes.length]);
 
   useEffect(() => {
-    dispatch(getTasks());
-    dispatch(getTaskTypes());
-    dispatch(getProducts());
-    dispatch(getHarvestUseTypes());
     dispatch(getAllUserFarmsByFarmId());
+    dispatch(getManagementPlansAndTasks());
```
Comment (antsgar) [inline packages/webapp/src/containers/Task/index.jsx:102]: I've tested it with this PR, and this is the change that fixes my current ticket -- with the previous code we were fetching the tasks first and the management plans at the end, here we're fetching all of the dependencies first and then the tasks

**Review by antsgar — APPROVED** [review]

**Comment by kathyavini** [comment]
> I am not sure about the requirements. It's probably better to move filter in tasksSelector
> 
> does this work?

It works, the `every` was the key!! 👌 I think we do need to filter on every task type that belongs to only a deleted management plan but works well that way too. Yes definitely better in this selector @JimmyRowland thank you!!! 🙏

**Review by JimmyRowland — APPROVED** [review]

**Pull request merged**